### ISSUE TITLE
GenServer process killed before ExUnit.on_exit callback

### ISSUE THREAD

**Issue by jazzyb** [elixir-lang/elixir#3279]
I am trying to stop a GenServer process within an `ExUnit.on_exit` callback:

```elixir
  setup do
    {:ok, pid} = TestProcess.start_link
    on_exit fn -> TestProcess.stop(pid) end
    {:ok, pid: pid}
  end

  test "process alive", context do
    assert Process.alive?(context[:pid])
  end
```

When the test is run, I get the following error even though the GenServer process is still alive in the test block:

```
[jazzyb@on_exit]$ mix test


  1) test process alive (OnExitTest)
     test/on_exit_test.exs:26
     ** (exit) exited in: GenServer.call(#PID<0.93.0>, :stop, 5000)
         ** (EXIT) no process
     stacktrace:
       (elixir) lib/gen_server.ex:356: GenServer.call/3
       (ex_unit) lib/ex_unit/on_exit_handler.ex:82: ExUnit.OnExitHandler.exec_callback/1
       (ex_unit) lib/ex_unit/on_exit_handler.ex:66: ExUnit.OnExitHandler.on_exit_runner_loop/0



Finished in 0.1 seconds (0.09s on load, 0.01s on tests)
1 tests, 1 failures

Randomized with seed 748275
```

My environment is Elixir 1.0.4, OTP 17.3.4, and FreeBSD 10.1:

```
[jazzyb@on_exit]$ uname -a
FreeBSD elixir.test.local 10.1-RELEASE-p9 FreeBSD 10.1-RELEASE-p9 #0: Tue Apr  7 01:09:46 UTC 2015     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  amd64
[jazzyb@on_exit]$ iex --version
Erlang/OTP 17 [erts-6.2.1] [source] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false]

Elixir 1.0.4
```

Pull my [on_exit repo](https://github.com/jazzyb/on_exit) for the reduced test case.  This issue is present in master.

**Comment by josevalim** [comment]
Because your TestProcess is linked to the actual test process and every test runs in its own process, once the test finishes, your TestProcess will exit due to link. This is not a bug, it is a feature. It means you don't need to explicitly stop it. :)

**Comment by jazzyb** [comment]
Sorry, I was certain that was a bug.  Thanks for the quick response!

**Comment by josevalim** [comment]
No problem! I jokingly said that "all bugs are now features" when we released 1.0. This one happens to actually be true :P
